File: examples/golang-push/rideshare/utility/utility.go
package utility

import (
	"context"
	"math/rand"
	"os"
	"time"

	"github.com/grafana/pyroscope-go"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"golang.org/x/sync/errgroup"
)

const (
	durationConstant = time.Duration(200 * time.Millisecond)
)

var (
	// vehicles contains all the known vehicle types.
	vehicles []Vehicle
)

func init() {
	// Populate the vehicles database.

	const (
		scooterCount = 100_000
		bikeCount    = 100_000
		carCount     = 50_000
	)

	vehicles = make([]Vehicle, 0, scooterCount+bikeCount+carCount)
	for i := 0; i < scooterCount; i++ {
		vehicles = append(vehicles, Vehicle{
			ID:   int(rand.Int()),
			Type: "scooter",
		})
	}
	for i := 0; i < bikeCount; i++ {
		vehicles = append(vehicles, Vehicle{
			ID:   int(rand.Int()),
			Type: "bike",
		})
	}
	for i := 0; i < carCount; i++ {
		vehicles = append(vehicles, Vehicle{
			ID:   int(rand.Int()),
			Type: "car",
		})
	}

	// Randomly order to the vehicles.
	rand.Shuffle(len(vehicles), func(i, j int) {
		vehicles[i], vehicles[j] = vehicles[j], vehicles[i]
	})
}

type Vehicle struct {
	ID   int
	Type string
}

func mutexLock(n int64) {
	var i int64 = 0

	// start time is number of seconds since epoch
	startTime := time.Now()

	// This changes the amplitude of cpu bars
	for time.Since(startTime) < time.Duration(n*30)*durationConstant {
		i++
	}
}

func checkDriverAvailability(n int64) {
	var i int64 = 0

	// start time is number of seconds since epoch
	startTime := time.Now()

	for time.Since(startTime) < time.Duration(n)*durationConstant {
		i++
	}

	// Every other minute this will artificially create make requests in eu-north region slow
	// this is just for demonstration purposes to show how performance impacts show up in the
	// flamegraph
	force_mutex_lock := time.Now().Minute()%2 == 0
	if os.Getenv("REGION") == "eu-north" && force_mutex_lock {
		mutexLock(n)
	}

}

func FindNearestVehicle(ctx context.Context, searchRadius int64, vehicle string) {
	ctx, span := otel.GetTracerProvider().Tracer("").Start(ctx, "FindNearestVehicle")
	span.SetAttributes(attribute.String("vehicle", vehicle))
	defer span.End()

	pyroscope.TagWrapper(ctx, pyroscope.Labels("vehicle", vehicle), func(ctx context.Context) {
		var i int64 = 0

		startTime := time.Now()
		for time.Since(startTime) < time.Duration(searchRadius)*durationConstant {
			i++
		}

		// Simulate finding an ID.
		ids := vehicleIDsFor(ctx, vehicle)
		_ = ids[rand.Intn(len(ids))]

		if vehicle == "car" {
			checkDriverAvailability(searchRadius)
		}
	})
}

func vehicleIDsFor(ctx context.Context, vehicleType string) []int {
	ctx, span := otel.GetTracerProvider().Tracer("").Start(ctx, "vehicleIDsFor")
	defer span.End()

	ids := make([]int, 0)
	for _, vehicle := range vehicles {
		if vehicle.Type != vehicleType {
			continue
		}
		ids = append(ids, vehicle.ID)
	}
	return ids
}

type RequestPool struct {
	stopAll chan chan struct{}
	group   *errgroup.Group
}

func (l *RequestPool) Handle(fn func() error) {
	done := make(chan struct{})

	// Leak a hanging goroutine.
	l.group.Go(func() error {
		// Run the function.
		err := fn()

		// Report the function is over.
		done <- struct{}{}

		force_leak := true
		if force_leak {
			// Report we leaked so we can eventually clean up and not
			// continuously crash the pod.
			stop := make(chan struct{})
			l.stopAll <- stop

			// Hang this goroutine until we're told to clean up.
			<-stop
		}
		return err
	})

	// Wait for function to finish.
	<-done
	close(done)
}

func (l *RequestPool) cleanUp() {
	ticker := time.NewTicker(1 * time.Minute)
	for {
		<-ticker.C
		if len(l.stopAll) < cap(l.stopAll) {
			// Don't clean up yet.
			continue
		}

		// Recover all hanging goroutines.
		for c := range l.stopAll {
			c <- struct{}{}
		}
	}
}

// NewRequestPool creates a goroutine pool of size n which can leak goroutines.
// Periodically it will clean up the leaked goroutines. Adjust n to leak more or
// less resources.
func NewRequestPool(n int) *RequestPool {
	pool := &RequestPool{
		stopAll: make(chan chan struct{}, n),
		group:   &errgroup.Group{},
	}
	pool.group.SetLimit(n)

	go pool.cleanUp()
	return pool
}

File: examples/language-sdk-instrumentation/python/rideshare/fastapi/lib/utility/utility.py
import time
import pyroscope
import os
from datetime import datetime

# How much time mutex_lock() takes relative to search_radius()
MUTEX_LOCK_MULTIPLIER = 2

# How much time check_driver_availability() takes relative to search_radius()
DRIVER_AVAILABILITY_MULTIPLIER = 0.5

ALLOCATION_SIZE = 64 * 1024
MAX_RETAINED_ALLOCATIONS = 256
RETAINED_ALLOCATIONS_AFTER_TRIM = 128
ALLOCATION_CHUNKS_BY_VEHICLE = {
    "bike": 1,
    "scooter": 2,
    "car": 4,
}
retained_allocations = []

def mutex_lock(n):
    i = 0
    start_time = time.time()
    while time.time() - start_time < n * MUTEX_LOCK_MULTIPLIER:
        i += 1

def check_driver_availability(n):
    i = 0
    start_time = time.time()
    while time.time() - start_time < n * DRIVER_AVAILABILITY_MULTIPLIER:
        i += 1

    # Every 4 minutes this will artificially create make requests in eu-north region slow
    # this is just for demonstration purposes to show how performance impacts show up in the
    # flamegraph

    force_mutex_lock = datetime.today().minute * 4 % 8 == 0
    if os.getenv("REGION") == "eu-north" and force_mutex_lock:
        mutex_lock(n)


def allocate_vehicle_memory(vehicle):
    for _ in range(ALLOCATION_CHUNKS_BY_VEHICLE[vehicle]):
        retained_allocations.append(bytearray(ALLOCATION_SIZE))
    if len(retained_allocations) >= MAX_RETAINED_ALLOCATIONS:
        del retained_allocations[:-RETAINED_ALLOCATIONS_AFTER_TRIM]


def find_nearest_vehicle(n, vehicle):
    with pyroscope.tag_wrapper({ "vehicle": vehicle}):
        i = 0
        start_time = time.time()
        while time.time() - start_time < n:
            i += 1
        allocate_vehicle_memory(vehicle)
        if vehicle == "car":
            check_driver_availability(n)
